File: app/src/main/java/com/lingion/sleepy/data/jw/XmuJw.kt
package com.lingion.sleepy.data.jw

import java.net.URLEncoder
import java.time.LocalDate

/**
 * 厦门大学教务(金智 gsapp/wdkbapp 微服务)登录与课表接口常量、纯函数。
 *
 * 端点与时序与仓库内 schedule/fetch_xmu_schedule.py 保持一致(该脚本已用真实账号
 * 验证过完整链路)。单学校版本下"学校"收敛为 [XMU_SCHOOL] 单一条目, 不再读
 * schools.json 目录。
 */
object XmuJw {

    /** 统一身份认证(ids) */
    const val IDP_BASE = "https://ids.xmu.edu.cn"

    /** 教务门户(jw) */
    const val JW_BASE = "https://jw.xmu.edu.cn"

    /**
     * 门户回调地址 — 登录成功跳回教务门户首页。保持原始未编码形态, 由调用方
     * (XmuAutoLoginClient)整体 encode 一次作为统一身份认证的 service 参数
     * (与 schedule/fetch_xmu_schedule.py 的 PORTAL_SERVICE 一致)。
     */
    val PORTAL_SERVICE: String =
        JW_BASE + "/login?service=" + JW_BASE + "/new/index.html"

    /** 教务登录入口(WebView 打开此地址; CAS 302 到统一身份认证) */
    const val LOGIN_URL = JW_BASE + "/login"

    /**
     * 统一使用的桌面版 Chrome UA。
     *
     * 厦大统一身份认证 / 教务门户按 UA 渲染: 手机 UA 会落到移动版页面, 与
     * schedule/fetch_xmu_schedule.py 验证过的桌面版接口时序不符 → 登录后抓取失败。
     * 手动登录 WebView 与自动登录客户端必须共用同一桌面 UA(与脚本一致)。
     */
    const val DESKTOP_UA = ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        + "(KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36")

    /** 课表微应用入口路径 — 抓课前必须先 GET 一次初始化应用会话, 否则数据接口 403 */
    const val APP_ENTRY_PATH = "/gsapp/sys/wdkbapp/*default/index.do?EMAP_LANG=zh&THEME=cherry"

    /** 金智 gsapp 微应用 API 基础路径 */
    const val GSAPP_API_BASE = JW_BASE + "/gsapp/sys/wdkbapp"

    /** 单学校版本的用户可见条目 — 替代原 181 所 schools.json 目录。 */
    val XMU_SCHOOL: JwSchoolInfo = JwSchoolInfo(
        sortKey = "X",
        name = "厦门大学",
        url = LOGIN_URL,
        type = JwProtocol.TYPE_XMU,
        aliases = listOf("xmu"),
        sortKeyFull = "xiamendaxue",
    )

    /**
     * 节次时段兜底表(厦门大学作息; wdkcb/queryXsskjc.do 未返回时使用)。
     * 与 schedule/fetch_xmu_schedule.py 的 _FALLBACK_PERIOD 同源。
     */
    val FALLBACK_PERIOD: Map<Int, Pair<String, String>> = mapOf(
        1 to ("08:00" to "08:45"),
        2 to ("08:55" to "09:40"),
        3 to ("10:10" to "10:55"),
        4 to ("11:05" to "11:50"),
        5 to ("14:30" to "15:15"),
        6 to ("15:25" to "16:10"),
        7 to ("16:40" to "17:25"),
        8 to ("17:35" to "18:20"),
        9 to ("19:10" to "19:55"),
        10 to ("20:05" to "20:50"),
        11 to ("21:00" to "21:45"),
    )

    fun encode(s: String): String = URLEncoder.encode(s, "UTF-8")

    /**
     * 按当前日期推断当前学年学期码: 9 月起为当年秋季("20261"), 其余月份回退到
     * 上一学年秋季。仅在接口学期列表拿不到时兜底使用。
     */
    fun inferSemesterCode(now: LocalDate): String? {
        val cur = if (now.monthValue >= 9) now.year else now.year - 1
        return if (cur in 1000..9999) "${cur}1" else null
    }
}

/**
 * 课表抓取统一载荷 — WebView 手动登录通道(XMU_FETCH_JS)与账号密码自动登录通道
 * ([XmuAutoLoginClient])共用, 交给 [JwImportViewModel.parseHtml] 解析后进入
 * 配置确认与落库流程。
 */
data class XmuFetchResult(
    /** wdkcb/queryXspkjg.do 的 JSON 全文(喂 [JwXmuParser]) */
    val courseJson: String,
    /** 节次时间字典 [(node, 开始 HH:mm, 结束 HH:mm)]; 为空时用厦大作息兜底 */
    val periods: List<Triple<Int, String, String>>,
    /** 学年学期码, 如 20261(供确认页展示, 不参与解析) */
    val semesterCode: String,
)

File: ios/XMUTimeTable/Services/XmuJw.swift
import Foundation

/// 厦大教务常量与纯函数 — 对齐 Android XmuJw
enum XmuJw {
    static let idpBase = "https://ids.xmu.edu.cn"
    static let jwBase = "https://jw.xmu.edu.cn"
    static let portalService = jwBase + "/login?service=" + jwBase + "/new/index.html"
    static let loginUrl = jwBase + "/login"
    static let desktopUA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
    static let appEntryPath = "/gsapp/sys/wdkbapp/*default/index.do?EMAP_LANG=zh&THEME=cherry"
    static let gsappApiBase = jwBase + "/gsapp/sys/wdkbapp"

    static let fallbackPeriod: [(Int, String, String)] = [
        (1, "08:00", "08:45"), (2, "08:55", "09:40"),
        (3, "10:10", "10:55"), (4, "11:05", "11:50"),
        (5, "14:30", "15:15"), (6, "15:25", "16:10"),
        (7, "16:40", "17:25"), (8, "17:35", "18:20"),
        (9, "19:10", "19:55"), (10, "20:05", "20:50"),
        (11, "21:00", "21:45")
    ]

    static func encode(_ s: String) -> String {
        s.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? s
    }

    static func inferSemesterCode(now: Date = Date()) -> String? {
        let cal = Calendar(identifier: .gregorian)
        let year = cal.component(.year, from: now)
        let month = cal.component(.month, from: now)
        let cur = month >= 9 ? year : year - 1
        if (1000...9999).contains(cur) { return "\(cur)1" }
        return nil
    }

    /// 解析金智课表 JSON 为 Course 数组
    static func parseCourseJson(
        _ json: String, tableId: Int64, timeJson: String, maxWeek: Int
    ) -> (courses: [Course], timeJson: String, maxWeek: Int, startDate: String)? {
        guard let data = json.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }

        var rows: [[String: Any]] = []
        if let datas = root["datas"] as? [String: Any] {
            for (_, v) in datas {
                if let d = v as? [String: Any], let r = d["rows"] as? [[String: Any]] {
                    rows = r
                    break
                }
            }
        }
        if rows.isEmpty, let r = root["rows"] as? [[String: Any]] {
            rows = r
        }
        if rows.isEmpty { return nil }

        var courses: [Course] = []
        var maxWeekSeen = 1
        var maxNodeSeen = 0
        var colorIdx = 0
        let palette = SleepyFormat.palette

        for row in rows {
            guard let name = (row["KCMC"] as? String ?? row["courseName"] as? String ?? row["kcmc"] as? String)?
                .trimmingCharacters(in: .whitespaces),
                  !name.isEmpty else { continue }

            let teacher = (row["JSXM"] as? String ?? row["teacher"] as? String ?? "") as String
            let room = (row["JASMC"] as? String ?? row["room"] as? String ?? "") as String
            let day = intOr(row["XQ"] ?? row["day"] ?? row["XQJ"], 1)
            let startNode = intOr(row["KSJC"] ?? row["startNode"] ?? row["start"], 1)
            let endNode = intOr(row["JSJC"] ?? row["endNode"] ?? row["end"], startNode)
            let startWeek = intOr(row["KSZ"] ?? row["startWeek"] ?? row["start"], 1)
            let endWeek = intOr(row["JSZ"] ?? row["endWeek"] ?? row["end"], 16)
            var weekType = intOr(row["SFJSMC"] ?? row["type"] ?? row["weekType"], 0)
            // 金智有时用文字描述单双周
            if let t = row["SFJSMC"] as? String {
                if t.contains("单") { weekType = 1 }
                else if t.contains("双") { weekType = 2 }
                else if t.contains("周") && !t.contains("单") && !t.contains("双") { weekType = 0 }
            }

            let step = max(1, endNode - startNode + 1)
            maxWeekSeen = max(maxWeekSeen, endWeek)
            maxNodeSeen = max(maxNodeSeen, endNode)

            colorIdx = (colorIdx % 9) + 1
            let color = palette[colorIdx] ?? SleepyFormat.autoColor

            courses.append(Course(
                id: 0,
                groupId: UUIDv5(namespace: name + "|" + teacher),
                tableId: tableId,
                courseName: name,
                teacher: teacher,
                room: room,
                note: "",
                day: max(1, min(7, day)),
                startNode: max(1, startNode),
                step: step,
                startWeek: max(1, startWeek),
                endWeek: max(endWeek, startWeek),
                type: weekType,
                color: color
            ))
        }

        return (courses, timeJson, max(maxWeek, maxWeekSeen), DateUtils.todayMondayString())
    }

    private static func intOr(_ any: Any?, _ def: Int) -> Int {
        if let i = any as? Int { return i }
        if let n = any as? NSNumber { return n.intValue }
        if let s = any as? String { return Int(s.trimmingCharacters(in: .whitespaces)) ?? def }
        return def
    }
}
